compiler/modules/types/list.bal

```diff
@@ -572,13 +572,42 @@ function listSubtypeIntersect(Context cx, SubtypeData t1, SubtypeData t2) return
 }
 
 function listSubtypeDiff(Context cx, SubtypeData t1, SubtypeData t2) returns SubtypeData {
+    if isDisjoint(cx, <Bdd> t1, <Bdd> t2) {
+        return t1;
+    }
     Bdd b = memoSubtypeDiff(cx.listMemo, <Bdd>t1, <Bdd>t2);
     if b != false && cx.listMemo[b] == () {
         //io:println("fresh BDD: ", bddToString(b));
     }
     return b;
 }
 
+function isDisjoint(Context cx, Bdd t1, Bdd t2) returns boolean {
+    if t1 !is BddNode || t2 !is BddNode {
+        return false;
+    }
+    if t1.left is true && t1.middle is false && t1.right is false &&
+       t2.left is true && t2.middle is false && t2.right is false {
+        ListAtomicType posAtom = cx.listAtomType(t1.atom);
+        ListAtomicType negAtom = cx.listAtomType(t2.atom);
+        // we only handle tuples
+        if !isNever(posAtom.rest) || !isNever(negAtom.rest) {
+            return false;
+        }
+        SemType[] posInitial = posAtom.members.initial; 
+        SemType[] negInitial = negAtom.members.initial; 
+        if posInitial.length() == negInitial.length() {
+            foreach int i in 0 ..< posInitial.length() {
+                if !isEmpty(cx, intersect(posInitial[i], negInitial[i])) {
```

Comment:
Not sure about using `isEmpty` here. Don't we want `isNever`?

Reply:
Is the logic right? The two lists are disjoint if any of the pairs of members are disjoint.

Reply:
fixed